From 7e9f4de74992af275f00a4bf9b0650acc6bb7ccf Mon Sep 17 00:00:00 2001
From: Takanari Hayama <taki@igel.co.jp>
Date: Thu, 13 Nov 2014 23:47:56 +0900
Subject: [PATCH 20/22] vsp-renderer: cope to negative offsets.

The new algorithm in v4l2-renderer may send negative offsets of
views.
---
 src/vsp-renderer.c |   19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/src/vsp-renderer.c b/src/vsp-renderer.c
index 1287d86..6be107c 100644
--- a/src/vsp-renderer.c
+++ b/src/vsp-renderer.c
@@ -921,9 +921,26 @@ vsp_comp_set_view(struct v4l2_renderer_device *dev, struct v4l2_surface_state *s
 		return -1;
 	}
 
-	DBG("set input %d (dmafd=%d): %dx%d@(%d,%d). alpha=%f\n",
+	if (vs->base.src_rect.width < 1 || vs->base.src_rect.height < 1) {
+		weston_log("ignoring the size of zeros < (%dx%d)\n", vs->base.src_rect.width, vs->base.src_rect.height);
+		return -1;
+	}
+
+	if (vs->base.src_rect.left < 0) {
+		vs->base.src_rect.width += vs->base.src_rect.left;
+		vs->base.src_rect.left = 0;
+	}
+
+	if (vs->base.src_rect.top < 0) {
+		vs->base.src_rect.height += vs->base.src_rect.top;
+		vs->base.src_rect.top = 0;
+	}
+
+	DBG("set input %d (dmafd=%d): %dx%d@(%d,%d)->%dx%d@(%d,%d). alpha=%f\n",
 	    vsp->input_count,
 	    vs->base.planes[0].dmafd,
+	    vs->base.src_rect.width, vs->base.src_rect.height,
+	    vs->base.src_rect.left, vs->base.src_rect.top,
 	    vs->base.dst_rect.width, vs->base.dst_rect.height,
 	    vs->base.dst_rect.left, vs->base.dst_rect.top,
 	    vs->base.alpha);
-- 
1.7.9.5

